### ISSUE TITLE
Consider upgrading to JUnit 5

### ISSUE THREAD

**Issue by mwoodiupui** [DSpace/DSpace#8219]
**Describe the bug**
JUnit 5 makes use of some handy Java 8+ features and adds some new capabilities.  JUnit 4 may not get as much attention (or any!) from developers of other testing tools.  We should consider whether it is worthwhile to upgrade.

There is a compatibility module supporting JUnit 4 tests, so we don't have to convert everything at once.

**Comment by tdonohue** [comment]
Agreed that we should move to JUnit 5.  I don't feel this is super high priority at the moment, so I'll pull this into Tier 2 board. But, if anyone is interested in starting this migration sooner, I'd support it.  Or, if we find a reason why we should migrate sooner rather than later we could re-prioritize.

Here's more on the JUnit 4 -> 5 Migration process: https://junit.org/junit5/docs/current/user-guide/#migrating-from-junit4

**Comment by Devamparikh** [comment]
Hey there, I have read some issues and discussions, you guys are planning to migrate Junit4 to Junit5, Can outsiders work on this issue? I want to help in migrating Junit4 to Junit5. Can I get some good first issues to work on? Thanks :)

**Comment by tdonohue** [comment]
@Devamparikh : Anyone is welcome to contribute code/features back to DSpace.  So, if this is of interest to you, we'd welcome it.  You'd want to work with the latest code on our `main` branch & submit a Pull Request once you have something working (at that point we'd review it & provide feedback). If you run into issues or have questions, you could ask them via comments in GitHub or use our Slack: https://wiki.lyrasis.org/display/DSPACE/Slack

That said, if this ticket is larger (it might be, I'm honestly not sure how difficult the migration from JUnit 4 -> 5 would be myself), you are also welcome to start with a smaller ticket.  We have a number of "good first issue" tickets already flagged:
* Backend (Java) "good first issue" tickets: https://github.com/DSpace/DSpace/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22
* Frontend (Angular) "good first issue" tickets: https://github.com/DSpace/dspace-angular/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22

In general though, anyone can contribute to DSpace at any time.  Nearly all developers are volunteers, and we don't have a centralized dev team (developers are from all over the world).  We just ask that if you want to work on a ticket, please mention it in the comments...that way we know not to try to find a different volunteer.

**Comment by Devamparikh** [comment]
@tdonohue I don't mind if this ticket is larger, I will work on this. Can you guide me along the process? What should be my first step for this? thank you :)

**Comment by tdonohue** [comment]
@Devamparikh : I don't have a step-by-step guide for updating from JUnit 4 to JUnit 5 (as I've never done it before).  But, I do know there are resources out there if you google such an upgrade, e.g. https://www.baeldung.com/junit-5-migration and https://blog.jetbrains.com/idea/2020/08/migrating-from-junit-4-to-junit-5/

That's probably the best place to start.  Upgrading to JUnit 5 may require fixing many/most of our test classes, so you'd likely need to get familiar with the test process: https://github.com/DSpace/DSpace/#testing

Beyond all that, once you have something to share, just create a Pull Request which links back to this ticket.  I'll then assign that PR to reviewers who can review your work & add comments/feedback if there's anything that needs updating.

Again, if this sounds too complex, you are also welcome to claim a smaller ticket & start with a small ticket just for practice.

**Comment by kshepherd** [comment]
I have started looking into this, and so far the biggest refactor job I've run into is the fact that in all the `assert*` methods, the message is now the last parameter, behind expected and actual, instead of the first parameter.
This isn't trivial (and is also not supported by the Intellij "Migrate JUnit 4 -> 5" migration tool), so I'm looking at what I could do with an OpenRewrite script or script(s)/tools that us an LSP to detect and work on those method usage refactors

**Comment by kshepherd** [comment]
@mwoodiupui @tdonohue I have started a draft PR with the results of my refactoring so far: https://github.com/DSpace/DSpace/pull/10764

**Comment by jesielviana** [comment]
I'm working on the migration — read more here: https://github.com/DSpace/DSpace/pull/10764#issuecomment-2978467735
